packages/packsquash/src/pack_file/commands_function_file.rs

```diff
@@ -0,0 +1,140 @@
+use std::borrow::Cow;
+
+use futures::{future, StreamExt};
+use thiserror::Error;
+use tokio::io::AsyncRead;
+use tokio_stream::Stream;
+use tokio_util::codec::{FramedRead, LinesCodec, LinesCodecError};
+
+use crate::config::CommandsFunctionFileOptions;
+use crate::pack_file::asset_type::PackFileAssetType;
+use crate::pack_file::util::{LineNumber, MarkLastDecorator};
+use crate::pack_file::AsyncReadAndSizeHint;
+
+use super::{OptimizedBytesChunk, PackFile, PackFileConstructor};
+
+#[cfg(test)]
+mod tests;
+
+/// Represents a line-oriented text file.
+pub struct CommandsFunctionFile<T: AsyncRead + Send + Unpin + 'static> {
+	read: T,
+	optimization_settings: CommandsFunctionFileOptions
+}
+
+/// Represents an error that may happen while optimizing commands function files.
+#[derive(Error, Debug)]
+pub enum OptimizationError {
+	#[error("Error while reading a line: {0}")]
+	TextLineRead(#[from] LinesCodecError),
+	#[error("Format error: Remove the leading slash in value of line {0}")]
+	RemoveLeadingSlash(LineNumber)
+}
+
+impl<T: AsyncRead + Send + Unpin + 'static> PackFile for CommandsFunctionFile<T> {
+	type ByteChunkType = Vec<u8>;
+	type OptimizationError = OptimizationError;
+	type OptimizedByteChunksStream =
+		impl Stream<Item = OptimizedBytesChunk<Self::ByteChunkType, Self::OptimizationError>>;
+
+	fn process(self) -> Self::OptimizedByteChunksStream {
+		let mut line_number = LineNumber::new();
+
+		let minify = self.optimization_settings.minify;
+		let strip_bom = self.optimization_settings.strip_bom;
+
+		MarkLastDecorator::new(FramedRead::new(
+			self.read,
+			// Limit line length to 16 KiB to bound memory consumption and be nice to Minecraft.
+			// Longer lines are probably an error, and will negatively affect Minecraft performance
+			LinesCodec::new_with_max_length(16 * 1024)
```

Comment:
Command parsing is done using [Brigadier's `StringReader` class](https://github.com/Mojang/brigadier/blob/cf754c4ef654160dca946889c11941634c5db3d5/src/main/java/com/mojang/brigadier/StringReader.java), which wraps a Java `String` with a `int` cursor to read substrings from it as needed. Therefore, my previous pessimistic upper bound turned to be the length limit for a single command read from a function, assuming that Minecraft has enough memory available to hold the string.

However, actually parsing the command involves additional memory allocations, and the vanilla launcher by default assigns a maximum Java heap size from 1 to 2 GiB. This means that the ~2 GiB upper limit can't be reached in practice with the default launcher options (and a 2 GiB command will likely be so slow to parse, too). Modded environments require more RAM, but it's fair to say that this extra RAM will be used by mods in practice, and thus it will not be available for command functions. A 16 MiB limit here looks sensible:

- Given that several files can be processed in parallel by PackSquash, and that the beefiest CPUs I know have 128 threads, this will bound PackSquash memory consumption to a maximum of 2 GiB, which should be a low memory consumption for sensible computers that use this kind of CPUs.
- With a maximum heap size of 2 GiB, this would mean that Minecraft can deal with up to 128 commands this long. 128 is a very low number, considering that functions tend to consist of tens of commands at least, and that other parts of the game and command parsing code require memory too (in fact, it's reasonable to say that the game requires at least 512 MiB for other purposes, which reduces the theoretical maximum number of commands to 96). I think it's extremely unlikely that someone ever needed a command 16 MiB long.
```suggestion
			// Limit line length to 16 MiB to bound memory consumption and be nice to Minecraft.
			// Longer lines are probably an error, and will negatively affect Minecraft performance
			LinesCodec::new_with_max_length(16 * 1024 * 1024)
```

Reply:
Perfect!! Thank you for committing :heart: